cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -1778,6 +1799,736 @@ class SortIndicesMetaFunction : public MetaFunction {
+      ArrayType& arr = *chunks_holder[chunks_holder.size() - 1];
+
+      std::vector<uint64_t> indices(arr.length());
+      uint64_t* indices_begin = indices.data();
+      uint64_t* indices_end = indices_begin + indices.size();
+      std::iota(indices_begin, indices_end, 0);
+
+      auto end_iter = PartitionNulls<ArrayType, NonStablePartitioner>(
+          indices_begin, indices_end, arr, 0);
+      auto kth_begin = indices_begin + options_.k;
+
+      if (kth_begin > end_iter) {
+        kth_begin = end_iter;
+      }
+      uint64_t* iter = indices_begin;
+      for (; iter != kth_begin && heap.size() < static_cast<size_t>(options_.k); ++iter) {
+        heap.Push(HeapItem{*iter, offset, &arr});
+      }
+      for (; iter != end_iter && !heap.empty(); ++iter) {
+        uint64_t x_index = *iter;
+        const auto& xval = GetView::LogicalValue(arr.GetView(x_index));
+        auto top_item = heap.top();
+        const auto& top_value =
+            GetView::LogicalValue(top_item.array->GetView(top_item.index));
+        if (comparator(xval, top_value)) {
+          heap.ReplaceTop(HeapItem{x_index, offset, &arr});
+        }
+      }
+      offset += chunk->length();
+    }
+
+    if (options_.keep_duplicates == true) {
+      offset = 0;
+      for (const auto& chunk : chunks_holder) {
+        ArrayType& arr = *chunk;
+
+        std::vector<uint64_t> indices(arr.length());
+        uint64_t* indices_begin = indices.data();
+        uint64_t* indices_end = indices_begin + indices.size();
+        std::iota(indices_begin, indices_end, 0);
+
+        auto iter = indices_begin;
+        for (; iter != indices_end; ++iter) {
+          uint64_t x_index = *iter;
+          auto top_item = heap.top();
+          if (x_index + offset != top_item.index + top_item.offset) {
+            const auto& xval = GetView::LogicalValue(arr.GetView(x_index));
+            const auto& top_value =
+                GetView::LogicalValue(top_item.array->GetView(top_item.index));
+            if (xval == top_value) {
+              heap.Push(HeapItem{x_index, offset, &arr});
+            }
+          }
+        }
+        offset += chunk->length();
+      }
+    }
+
+    int64_t out_size = static_cast<int64_t>(heap.size());
+    ARROW_ASSIGN_OR_RAISE(
+        auto take_indices,
+        MakeMutableArrayForFixedSizedType(uint64(), out_size, ctx_->memory_pool()));
+    auto* out_cbegin = take_indices->GetMutableValues<uint64_t>(1) + out_size - 1;
+    while (heap.size() > 0) {
+      auto top_item = heap.top();
+      *out_cbegin = top_item.index + top_item.offset;
+      heap.Pop();
+      --out_cbegin;
+    }
+    ARROW_ASSIGN_OR_RAISE(auto chunked_select_k,
+                          Take(Datum(chunked_array_), Datum(std::move(take_indices)),
+                               TakeOptions::NoBoundsCheck(), ctx_));
+    ARROW_ASSIGN_OR_RAISE(
+        auto select_k,
+        Concatenate(chunked_select_k.chunked_array()->chunks(), ctx_->memory_pool()));
```

Comment:
Why do we concatenate here? If the input is a chunked array I would expect the output to also be a chunked array.

Reply:
Update: I changed these APIs to return only indices like SortIndices.